### PR TITLE
Multiple dataset training Web Support

### LINKED ISSUE #391: [FEATURE] Allow training of multiple datasets at once
I'm not too sure of the use case, @mattdawkins please comment the specifics.

### PR DESCRIPTION
**Pull request by BryonLewis**
Fixes #391 

**NOTE** - Need the latest `kitware/viame:gpu-algorithms-latest` for the input_list  to work properly.

- Enabled Training menu button when one **or more** items are selected in a folder
- Changes the input type for training to an array of `folderIds` and updates the API in the relevant locations.
- Server now takes in the JSON data for the folderIds and does a preprocess check on each dataset to ensure there is some groundtruth csv files there.
- Updated the training to remove the labels.txt and use the new `-il input_folder_list.txt` and the `-it input_groundtruth_list.txt` for specifiying the data.
- Since it still requires a folder structure for orangization I kept the `organize_folder_for_training` but removed the `labels.txt` stuff.
- `--no-query` is added to the groundtruth command so it will use all types that would be in the l`abels.txt` by default and prevent the user from being prompted to accept.

I've tested by taking to small datasets with different track types in it and training on them.  Then I would run the trained model on another small dataset and ensure that it is using types from both datasets.

Additionally I trained across different folders by using the `/viame/train` endpoint and manually specifying folderIds across different root folders and different public users.  It trained successfully and the resulting pipeline incorporated types across the different folders.

### CODE REVIEW

**Comment by subdavis** [comment]
> Since it still requires a folder structure for orangization

Could you explain this part?  I didn't expect that anything would need to move.  You could just do a simple download of each dataset from Girder, then point to the data in-place without moving anything.

Also, I expect you'd like to merge this before #487.  I'm fine with that.  Just to confirm, this will work with arbitrary dataset ids right?  They don't have to be siblings?

**Comment by BryonLewis** [comment]
> Could you explain this part? I didn't expect that anything would need to move. You could just do a simple download of each dataset from Girder, then point to the data in-place without moving anything.

besides my massive spelling mistake there (orangization).  Bad choice of words for the explanation.  I had the assumption that the testing of the ground_truth to see if it is a directory was in there because of some legacy items where the `meta.detection` might provide a directory instead of a csv so that test needed to remain in there.  Really all `organize_folder_for_training` does right now is check if the ground_truth is a directory, if it is it will copy the first .csv out of it into the training data directory associated with that groundtruth and then deletes the folder.  If it is a file it just renames it to ground_truth.csv which is unnecessary but cleaner.   I could keep all ground truth at the root level, but if I'm doing the test and possibly moving it why not keep it a bit more organized.  I should probably clean up the description and what it is called.



> Also, I expect you'd like to merge this before #487. I'm fine with that. Just to confirm, this will work with arbitrary dataset ids right? They don't have to be siblings?

Yeah that was the second part of my testing, I was training across different user's public folders, just required that I manually create the array of dataset ids and call the endpoint because there is no UI  for it currently.

File: server/viame_tasks/tasks.py [hunk 1]
```diff
@@ -184,64 +184,87 @@ def run_pipeline(self: Task, params: PipelineJob):
...
+    source_folder_list: List[Dict],
+    groundtruth_list: List[Dict],
     pipeline_name: str,
     config: str,
 ):
     """
     Train a pipeline by making a call to viame_train_detector
 
-    :param source_folder: The Girder Folder to pull training data from
+    :param source_folder_list: The Girder Folder to pull training data from
```
Comment (AlmightyYakob) [inline server/viame_tasks/tasks.py:195]:
```suggestion
    :param source_folder_list: The Girder Folders to pull training data from
```

File: server/viame_tasks/tasks.py [hunk 2]
```diff
@@ -184,64 +184,87 @@ def run_pipeline(self: Task, params: PipelineJob):
...
     :param results_folder: The Girder Folder to place the results of training into
-    :param groundtruth: The relative path to either the file containing detections,
+    :param groundtruth_list: The relative path to either the file containing detections,
         or the folder containing that file.
```
Comment (AlmightyYakob) [inline server/viame_tasks/tasks.py:198]:
```suggestion
    :param groundtruth_list: A list of relative paths to either a file containing detections,
        or a folder containing that file.
```

File: server/viame_tasks/tasks.py [hunk 3]
```diff
@@ -184,64 +184,87 @@ def run_pipeline(self: Task, params: PipelineJob):
...
-        for item in training_data:
-            gc.downloadItem(str(item["_id"]), download_path)
+        # Generator of items
+        for index in range(len(source_folder_list)):
+            source_folder = source_folder_list[index]
+            groundtruth = groundtruth_list[index]
+            download_path = Path(tempfile.mkdtemp(dir=root_data_dir))
+            training_data = gc.listItem(source_folder["_id"])
+            trained_on_list.append(str(source_folder["_id"]))
+
```
Comment (AlmightyYakob) [inline server/viame_tasks/tasks.py:229]: Since the for loop isn't actually a generator but `gc.listItem` is:
```suggestion
        for index in range(len(source_folder_list)):
            source_folder = source_folder_list[index]
            groundtruth = groundtruth_list[index]
            download_path = Path(tempfile.mkdtemp(dir=root_data_dir)
            trained_on_list.append(str(source_folder["_id"]))

            # Generator of items
            training_data = gc.listItem(source_folder["_id"])

```

File: server/viame_tasks/tasks.py [hunk 4]
```diff
@@ -184,64 +184,87 @@ def run_pipeline(self: Task, params: PipelineJob):
...
     training_executable = viame_install_path / "bin" / "viame_train_detector"
     config_file = pipeline_base_path / config
 
     pipeline_name = pipeline_name.replace(" ", "_")
 
+    if len(source_folder_list) != len(groundtruth_list):
+        raise Exception("Ground truth doesn't exist for all folders")
+
+    trained_on_list = []
+    input_groundtruth_list = []
```
Comment (AlmightyYakob) [inline server/viame_tasks/tasks.py:216]: Type hints / comments would be useful here

**Review by AlmightyYakob — COMMENTED** [review]
Just some minor things but it looks good, haven't tested locally yet

**Comment by AlmightyYakob** [comment]
> Really all `organize_folder_for_training` does right now is check if the ground_truth is a directory, if it is it will copy the first .csv out of it into the training data directory associated with that groundtruth and then deletes the folder. If it is a file it just renames it to ground_truth.csv which is unnecessary but cleaner.

For the record, the reason this is done is in case the girder item has multiple files. If it does, it's a folder when downloaded. Otherwise it's just a file. Since we still use the `csv_detection_file` method to ensure a csv file on the item, this is likely always the case.

File: server/viame_server/viame.py [hunk 5]
```diff
@@ -129,20 +126,33 @@ def run_pipeline_task(self, folder, pipeline: PipelineDescription):
...
+
+        for folderId in folderIds:
+            folder = Folder().load(folderId, level=AccessType.READ, user=user)
+            if folder is None:
+                raise Exception(f"Cannot access folder {folderId}")
+            folder_names.append(folder['name'])
+            detections = list(
+                Item().find({"meta.detection": str(folderId)}).sort([("created", -1)])
+            )
+            detection = detections[0] if detections else None
```
Comment (subdavis) [inline server/viame_server/viame.py:147]: Use `_load_detections()` helper function.

Comment (subdavis) [inline server/viame_server/viame.py:147]: Maybe refactor `viame_detection.py` `_load_detections()` helper function?

Comment (subdavis) [reply server/viame_server/viame.py:147]: Eh, this can be done later.

**Review by subdavis — APPROVED** [review]
:+1:
